woocommerce/woocommerce
# Fix analytics chart previous year data when date range includes 29 Feb

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/47497

I tried to fix leap year data in general, however, it turns out analytics code, dates wrangling, and chart rendering are too tightly tied together for me to be able to confidently fix them without making a major revamp.

This PR attempts to patch the bug by "bumping up" the previous year's data starting from 29th Feb when:

1. Charts comparison set to by day
1. The current date is 29th February
2. The previous date is 1st March

This ensures the fix is narrowly targeted to this specific bug and least likely to introduce unintended side effects.

### How to test the changes in this Pull Request:

1. Download and install [wc-smooth-generator](https://github.com/woocommerce/wc-smooth-generator/releases/tag/1.1.0) plugin
2. Run the following commands in wp-cli: <p>
```
wp wc generate orders 1 --date-start=2020-02-29 --date-end=2020-02-29 --status=completed &&
wp wc generate orders 1 --date-start=2019-02-28 --date-end=2019-02-28 --status=completed &&
wp wc generate orders 1 --date-start=2019-03-01 --date-end=2019-03-01 --status=completed
```

3. Go to `Analytics > Settings` and set `Date type` to `Date created` and save
4. Go to `Analytics > Orders` and set `Date range` to custom tab, `02/01/2020` to `03/31/2020` and update
5. Ensure comparison is set to `By day` ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/9c57d1ce-5601-4ba7-aa0c-59704441bae8))
7. Hover over 28th February and observe the previous year label is `Invalid date` and the value is `0` ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/bd6672f7-d3e9-4807-bd99-d3bf8590f3e3))
8. Hover over 1st March and observe there's 1 order for `March 1 2019` ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/9aee87cd-42c4-4249-947e-718ade69b4e6))

### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
